File: services/notification/src/clients/secretsLoader.js
import { SecretsManagerClient, GetSecretValueCommand } from "@aws-sdk/client-secrets-manager";

/**
 * Loads secrets from AWS Secrets Manager into `process.env`, so every existing
 * (synchronous) config path keeps working untouched.
 *
 * Two modes, the same shape as the other switches (`KAFKA_AUTH`, `WS_PROVIDER`,
 * `OPENSEARCH_AUTH`):
 *
 *   env    (default, incl. unset/blank) → no-op; values come from .env as they
 *                                         always have
 *   awssm  → fetch each spec and populate the mapped env vars before anything
 *            reads them
 *
 * WHY hydrate `process.env` instead of returning values: config.js and every
 * client are synchronous, and a secret fetch is not. Hydrating once at startup
 * keeps the async boundary in exactly one place — the adapter — rather than
 * turning `getDependencies()` and its callers async for four values. It also
 * means the local path is literally unchanged code.
 *
 * Fetched values OVERRIDE anything already in the environment. A deployment that
 * says "get secrets from Secrets Manager" must not silently keep using a stale
 * value baked into a task definition.
 *
 * DELIBERATELY NOT CACHED ACROSS INVOCATIONS beyond the process's own env: this
 * runs once per cold start, and CLAUDE.md forbids depending on warm-start reuse
 * for correctness. The flip side is rotation — a rotated secret is picked up on
 * the next cold start, not mid-life. For a long-lived Fargate task that means a
 * restart. Fine for these four (none rotate automatically today); revisit if a
 * rotation schedule is ever attached.
 */

/**
 * @param {Array<{secretId: string, map: Record<string,string>, whole?: string, required?: boolean}>} specs
 *   secretId — Secrets Manager name or ARN
 *   map      — JSON key → env var name
 *   whole    — env var that receives the ENTIRE secret string (for opaque blobs
 *              like a service-account JSON, which must not be destructured)
 *   required — throw if missing (default false: log and fall through to .env)
 */
export async function hydrateSecrets(specs) {
  const mode = (process.env.SECRETS_PROVIDER ?? "").trim().toLowerCase();
  if (mode === "" || mode === "env") {
    return { provider: "env", loaded: [] };
  }
  if (mode !== "awssm") {
    throw new Error(`SECRETS_PROVIDER must be 'env' or 'awssm', got: '${process.env.SECRETS_PROVIDER}'`);
  }

  const client = new SecretsManagerClient({
    region: process.env.AWS_REGION,
    // LocalStack in dev; unset in AWS so the SDK uses the real endpoint.
    ...(process.env.SECRETS_MANAGER_ENDPOINT
      ? { endpoint: process.env.SECRETS_MANAGER_ENDPOINT }
      : {}),
  });

  const loaded = [];
  for (const spec of specs) {
    try {
      const res = await client.send(new GetSecretValueCommand({ SecretId: spec.secretId }));
      const raw = res.SecretString;
      if (!raw) {
        throw new Error("secret has no SecretString (binary secrets are not supported)");
      }

      if (spec.whole) {
        process.env[spec.whole] = raw;
        loaded.push(spec.secretId);
        continue;
      }

      const parsed = JSON.parse(raw);
      for (const [jsonKey, envVar] of Object.entries(spec.map ?? {})) {
        if (parsed[jsonKey] === undefined || parsed[jsonKey] === null) {
          // eslint-disable-next-line no-console
          console.warn(`[secrets] ${spec.secretId} has no '${jsonKey}' — leaving ${envVar} as-is`);
          continue;
        }
        process.env[envVar] = String(parsed[jsonKey]);
      }
      loaded.push(spec.secretId);
    } catch (err) {
      if (spec.required) {
        // A required secret that can't be read is fatal: starting up and failing
        // later, per request, is strictly worse than not starting.
        throw new Error(`[secrets] required secret ${spec.secretId} unavailable: ${err.message}`);
      }
      // eslint-disable-next-line no-console
      console.warn(
        `[secrets] ${spec.secretId} unavailable (${err.message}) — falling back to the environment`
      );
    }
  }
  return { provider: "awssm", loaded };
}

File: services/presence-connection/src/main/java/dev/rstrickland/chat/presence/clients/SecretsLoader.java
package dev.rstrickland.chat.presence.clients;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.net.URI;
import java.util.Locale;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.secretsmanager.SecretsManagerClient;
import software.amazon.awssdk.services.secretsmanager.model.GetSecretValueRequest;

/**
 * Reads secrets from AWS Secrets Manager, with the environment as the fallback.
 * The Java counterpart to the Node services' `clients/secretsLoader.js`, driven
 * by the SAME {@code SECRETS_PROVIDER} switch:
 *
 * <ul>
 *   <li><b>env</b> (default, incl. unset/blank) — return the env value unchanged;
 *       local dev is untouched.
 *   <li><b>awssm</b> — fetch the secret and use the named JSON field, falling
 *       back to the env value if it can't be read.
 * </ul>
 *
 * Falling back rather than failing is deliberate: this service VERIFIES the
 * internal API key on its `/internal/presence/...` route. If the value can't be
 * loaded, the env value still guards that one route while every bearer-authed
 * route and the WebSocket handlers keep working — a narrower failure than
 * refusing to start.
 *
 * No caching beyond the value Config holds for the process's lifetime: this runs
 * once at startup, and CLAUDE.md forbids depending on warm-start reuse for
 * correctness. The trade-off is rotation — a rotated secret needs a restart.
 *
 * Duplicated (bar the package) in the messaging service, like KafkaSecurity, for
 * the same reason: there is no shared Java module in this repo.
 */
public final class SecretsLoader {

  public static final String ENV_VAR = "SECRETS_PROVIDER";

  private static final ObjectMapper MAPPER = new ObjectMapper();

  private SecretsLoader() {}

  /**
   * @param secretId Secrets Manager name or ARN
   * @param jsonField field to read out of the secret's JSON document
   * @param envFallback value from the environment, used when the provider is
   *     `env` or the fetch fails
   * @return the resolved value, or {@code envFallback}
   */
  public static String resolve(String secretId, String jsonField, String envFallback) {
    String raw = System.getenv(ENV_VAR);
    String mode = (raw == null || raw.isBlank()) ? "env" : raw.trim().toLowerCase(Locale.ROOT);

    if ("env".equals(mode)) {
      return envFallback;
    }
    if (!"awssm".equals(mode)) {
      // Fail fast: an unrecognised provider is a deployment mistake, and silently
      // using .env values in AWS is exactly what this feature exists to stop.
      throw new IllegalStateException(ENV_VAR + " must be 'env' or 'awssm', got: '" + raw + "'");
    }

    try (SecretsManagerClient client = buildClient()) {
      String json =
          client
              .getSecretValue(GetSecretValueRequest.builder().secretId(secretId).build())
              .secretString();
      JsonNode node = MAPPER.readTree(json).get(jsonField);
      if (node == null || node.isNull()) {
        System.err.println(
            "[secrets] " + secretId + " has no '" + jsonField + "' — using the environment value");
        return envFallback;
      }
      System.out.println("[secrets] loaded " + secretId + " from Secrets Manager");
      return node.asText();
    } catch (Exception e) {
      System.err.println(
          "[secrets] " + secretId + " unavailable (" + e.getMessage()
              + ") — falling back to the environment");
      return envFallback;
    }
  }

  private static SecretsManagerClient buildClient() {
    var builder =
        SecretsManagerClient.builder()
            .region(Region.of(envOr("AWS_REGION", "us-east-1")));
    // LocalStack in dev; unset in AWS so the SDK resolves the real endpoint.
    String endpoint = System.getenv("SECRETS_MANAGER_ENDPOINT");
    if (endpoint != null && !endpoint.isBlank()) {
      builder.endpointOverride(URI.create(endpoint));
    }
    return builder.build();
  }

  private static String envOr(String name, String fallback) {
    String v = System.getenv(name);
    return (v == null || v.isBlank()) ? fallback : v;
  }
}
